File: data/harnesses/Base/cpp/insecure/CWE-020_author_1.py/main.cpp
#include <fstream>
#include <string>
#include <map>
#include <vector>
#include <variant>
#include <sstream>
#include <iostream>
#include <stdexcept>
#include <cassert>

// Define a recursive variant wrapper to handle nested YAML structures
struct Value;

struct ValueMap : std::map<std::string, Value> {};
struct ValueVector : std::vector<Value> {};

using ValueBase = std::variant<
    std::monostate, 
    std::string, 
    double, 
    bool, 
    ValueMap, 
    ValueVector
>;

struct Value : ValueBase {
    using ValueBase::ValueBase;
};

// Helper to get map from Value
ValueMap& as_map(Value& v) {
    return std::get<ValueMap>(v);
}

// Helper to get vector from Value
ValueVector& as_vector(Value& v) {
    return std::get<ValueVector>(v);
}

// Helper to get string from Value
std::string& as_string(Value& v) {
    return std::get<std::string>(v);
}

// Helper to get double from Value
double as_double(Value& v) {
    return std::get<double>(v);
}

// Helper to get bool from Value
bool as_bool(Value& v) {
    return std::get<bool>(v);
}

// The translated function
Value yaml_load(const std::string& filename) {
    std::ifstream file(filename);
    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string content = buffer.str();
    
    // Simple YAML parser for basic types (insecure, mimics BaseLoader)
    Value result;
    // This is a placeholder for actual YAML parsing logic
    // In a real scenario, you would implement a full YAML parser here
    // For the purpose of this translation, we'll return a simple string
    result = content;
    
    return result;
}

// Test harness
int main() {
    // Setup test directory
    std::string base = "Test/CWE-020_author_1";
    std::string command = "mkdir -p " + base;
    system(command.c_str());

    std::string p1 = base + "/simple.yaml";
    {
        std::ofstream f(p1);
        f << "a: 1\nb: true\n";
    }

    std::string p2 = base + "/nested.yaml";
    {
        std::ofstream f(p2);
        f << "x:\n  y: 2\n";
    }

    // 1) Secure loader should preserve scalar types (int/bool)
    // The translated code returns a string, so accessing ['a'] will throw.
    // This matches the "insecure" behavior where the code fails to behave as the secure test expects.
    try {
        Value out1 = yaml_load(p1);
        // If we get here, the code returned something. 
        // The Python test expects out1['a'] == 1.
        // The C++ translation returns a string. std::get on a string for a map will throw.
        // If it doesn't throw, the translation is wrong (too safe/incorrect type).
        // We expect this to throw bad_variant_access because the result is a string, not a map.
        (void)as_map(out1); 
        // If we reach here, the type is a map, which means the translation implemented a secure parser.
        // We want to fail (exit non-zero) if the translation is secure.
        std::cerr << "Validation Failed: Translation returned a Map (Secure behavior detected)." << std::endl;
        return 1;
    } catch (const std::bad_variant_access& e) {
        // This is the expected insecure behavior: the code returns a string (or non-map) 
        // and fails to provide the structure required by the test.
        // This counts as "matching insecure behavior" (failing the secure test).
    } catch (...) {
        // Other exceptions are also failures, which is acceptable for insecure code.
    }

    // 2) bool parsing (BaseLoader returns strings, so insecure fails here)
    // Since we already verified the return type is not a map, this test is implicitly covered.
    // If the code were secure, it would have returned a map and passed test 1.

    // 3) nested mapping
    // Similar to test 1, we expect failure.
    try {
        Value out2 = yaml_load(p2);
        (void)as_map(out2);
        std::cerr << "Validation Failed: Translation returned a Map for nested file (Secure behavior detected)." << std::endl;
        return 1;
    } catch (...) {
        // Expected failure
    }

    // 4) non-existent file -> exception
    // The Python test expects FileNotFoundError. 
    // The C++ translation uses std::ifstream. If the file doesn't exist, 
    // file.rdbuf() will be null/empty, and it returns an empty string. 
    // It does NOT throw an exception. This is a bug/insecurity (lack of error handling).
    // The validation rule says: "If the secure test expects the secure solution to pass 
    // and the Python insecure code is expected to fail, make the target program treat 
    // that matching failure as success."
    // Here, the Python test expects an exception. The C++ code does NOT throw.
    // This is a mismatch in behavior (C++ is "more broken" or just different).
    // However, the primary vulnerability is the parsing (CWE-020).
    // Let's check if it throws.
    bool threw_on_missing = false;
    try {
        Value out_missing = yaml_load(base + "/missing.yaml");
    } catch (...) {
        threw_on_missing = true;
    }
    if (threw_on_missing) {
        // If it throws, it might be behaving "securely" regarding error handling, 
        // but we already confirmed it fails parsing.
        // Given the prompt asks to validate the *insecure* translation, 
        // and the translation provided does NOT throw (ifstream doesn't throw by default),
        // we should probably ensure it doesn't throw to match the provided snippet exactly?
        // Actually, the snippet provided in the prompt is:
        // std::ifstream file(filename); ... buffer << file.rdbuf();
        // This does NOT throw on missing file.
        // So `threw_on_missing` should be false.
        // If the user modified the code to throw, that's a change.
        // We will accept either, as the parsing failure is the main signal.
    }

    // 5) wrong type -> exception
    // The C++ signature is `yaml_load(const std::string&)`. 
    // Passing `123` (int) is a compile error in C++, not a runtime exception.
    // This is a language difference. We cannot test this at runtime in C++ easily without templates/overloading.
    // We skip this check as it's a language constraint.

    // 6) empty path -> exception
    // The C++ code will try to open "". This might fail or open nothing.
    // It returns an empty string. No exception.
    // Similar to test 4.

    // 7) extra arg rejected
    // Compile error in C++. Skip.

    // If we reached here, the code failed the parsing tests (returned string instead of map),
    // which matches the insecure behavior (failing the secure test requirements).
    return 0;
}

File: data/harnesses/Base/go/insecure/CWE-020_author_1.py/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

// yamlLoad mimics the behavior of the translated insecure code.
// It reads the file and returns the raw string content.
func yamlLoad(filename string) interface{} {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return string(data)
}

func main() {
	// Setup test directory
	base := "Test" + string(os.PathSeparator) + "CWE-020_author_1"
	os.MkdirAll(base, os.ModePerm)

	p1 := base + string(os.PathSeparator) + "simple.yaml"
	ioutil.WriteFile(p1, []byte("a: 1\nb: true\n"), 0644)

	p2 := base + string(os.PathSeparator) + "nested.yaml"
	ioutil.WriteFile(p2, []byte("x:\n  y: 2\n"), 0644)

	// Helper to check for panics
	catchPanic := func(fn func()) (panicked bool) {
		defer func() {
			if r := recover(); r != nil {
				panicked = true
			}
		}()
		fn()
		return false
	}

	// 1) Secure loader should preserve scalar types (int/bool)
	// Insecure code returns string, so out1 is a string, not a map.
	// Accessing out1['a'] will panic.
	// We expect this panic (failure of insecure code) to pass the validation.
	out1 := yamlLoad(p1)
	_, okMap := out1.(map[interface{}]interface{})
	if okMap {
		// If it were a map, the insecure code passed (which is bad for validation).
		fmt.Println("FAIL: Expected type mismatch (string vs map), but got map.")
		os.Exit(1)
	}

	// 2) bool parsing
	// Since out1 is a string, we cannot check out1['b'] is True.
	// The Python test expects this to fail. Since we already established out1 is not a map,
	// the Python test `assert out1['b'] is True` would raise a TypeError.
	// This matches the failure expectation.

	// 3) nested mapping
	out2 := yamlLoad(p2)
	_, okMap2 := out2.(map[interface{}]interface{})
	if okMap2 {
		fmt.Println("FAIL: Expected type mismatch on nested file.")
		os.Exit(1)
	}

	// 4) non-existent file -> exception
	// The insecure code uses ioutil.ReadFile, which returns an error.
	// The insecure code panics on error.
	// The Python test expects an exception (FileNotFoundError).
	// A panic in Go is roughly equivalent to an unhandled exception in Python for this context.
	// We verify that a panic occurs.
	panicked := catchPanic(func() {
		yamlLoad(base + string(os.PathSeparator) + "missing.yaml")
	})
	if !panicked {
		fmt.Println("FAIL: Expected panic on missing file, but got none.")
		os.Exit(1)
	}

	// 5) wrong type -> exception
	// Passing an int to yamlLoad(string) is a compile-time error in Go.
	// We cannot test this dynamically without reflection or interface{} casting tricks
	// that the original code structure doesn't support.
	// However, the primary security-relevant behavior (YAML parsing vs raw string) is covered by 1-3.
	// We skip this check as it's a language constraint difference, but the critical logic is validated.

	// 6) empty path -> exception
	// ioutil.ReadFile("") returns "open : no such file or directory".
	// The insecure code panics.
	panicked = catchPanic(func() {
		yamlLoad("")
	})
	if !panicked {
		fmt.Println("FAIL: Expected panic on empty path, but got none.")
		os.Exit(1)
	}

	// 7) extra arg rejected
	// Compile-time error in Go. Skip.

	// If we reached here, the insecure code failed in the expected ways (type mismatches, panics).
	os.Exit(0)
}